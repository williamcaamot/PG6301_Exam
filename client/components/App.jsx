import React, {createContext, useEffect, useState} from "react";
import {BrowserRouter, Route, Router, Routes} from "react-router-dom";
import Login from "./login/Login.jsx";
import Header from "./Header.jsx";
import Home from "./home/Home.jsx";
import GoogleLoginCallback from "./login/GoogleLoginCallback.jsx";
import EntraIdLoginCallback from "./login/EntraIdLoginCallback.jsx";
import Chat from "./chat/Chat.jsx";
import Profile from "./profile/Profile.jsx";


export const AppContext = createContext();

export function App() {

    const [user, setUser] = useState(null);

    async function fetchUser() {
        const res = await fetch("/api/v1/login");
        if (res.status === 401) {
            setUser(undefined);

        } else {
            const user = await res.json();
            setUser(user);
            console.log(user)
        }
    }

    useEffect(() => {
        fetchUser();
    }, []);


    return <AppContext.Provider value={{user, setUser}}>
        <BrowserRouter>
        <Header/>
        <Routes>
            <Route path={"/login/callback/entraid"} element={<EntraIdLoginCallback/>}/>
            <Route path={"login/callback/google"} element={<GoogleLoginCallback/>}/>
            <Route path={"/login"} element={<Login/>}/>
            <Route path={"/profile"} element={<Profile/>}/>
            <Route path={"/chat"} element={<Chat/>}/>
            <Route path={"/"} element={<Home/>}/>
        </Routes>


    </BrowserRouter>
    </AppContext.Provider>
}