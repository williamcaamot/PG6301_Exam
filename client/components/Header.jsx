import React, {useContext} from "react";
import {Link} from "react-router-dom";
import {AppContext} from "./App.jsx";

function Header() {

    const { user } = useContext(AppContext);

    return <>
        <div className={"flexWrapper"}>
            <div className={"headerWrapper"}>
                <div>
                    <h1>Banter Battle - Chat website</h1>
                </div>
                <div>
                    <nav>
                        <Link to="/" className={"mainMenuLink"}>Hjem</Link>
                        <Link to="/chat" className={"mainMenuLink"}>Chat</Link>
                        <Link to="/chatroom/add" className={"mainMenuLink"}>Add new chat room</Link>

                    </nav>
                </div>
                <div style={{flexDirection:"row-reverse"}}>
                    {user ?
                        <Link to="/profile" className={"mainMenuLink"}>Velkommen, {user.nickname}</Link>
                        :
                        <Link to="/login" className={"mainMenuLink"}>Logg inn</Link>
                    }

                </div>
            </div>
        </div>
    </>
}

export default Header;