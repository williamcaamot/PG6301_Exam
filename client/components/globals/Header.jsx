import React, {useContext} from "react";
import {Link} from "react-router-dom";
import {AppContext} from "../App.jsx";

function Header() {

    const { user } = useContext(AppContext);

    return <>
        <div className={"flexWrapper"}>
            <div className={"headerWrapper"}>
                <div style={{padding:"30px"}}>
                    <h1>ChatMania</h1>
                </div>
                <div>
                    <nav>
                        <Link to="/" className={"mainMenuLink"}>Hjem</Link>
                        <Link to="/chat" className={"mainMenuLink"}>Chat</Link>
                        <Link to="/chatroom/add" className={"mainMenuLink"}>Add new chat room</Link>
                        {user ?
                            <Link to="/profile" className={"mainMenuLink"}>Velkommen, {user.nickname}</Link>
                            :
                            <Link to="/login" className={"mainMenuLink"}>Logg inn</Link>
                        }
                    </nav>
                </div>
            </div>
        </div>
    </>
}

export default Header;