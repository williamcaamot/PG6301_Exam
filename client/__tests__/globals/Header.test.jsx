import renderer, {act} from "react-test-renderer";
import {MemoryRouter} from "react-router-dom";
import Header from "../../components/globals/Header.jsx";
import {AppContext} from "../../components/App.jsx";

describe("header component", () => {

    it("display the header with user information", () => {

        const user = {
            message: "daww",
            sender: "williamcoucheronaamot@gmail.com",
            nickname: "WILLIAM123",
            picture: "https://lh3.googleusercontent.com/a/ACg8ocIOgSmXkWTJqoKPe5Hk3JlkkhfUNer4sWkVo3UuIa_dzg=s96-c",
            time: "2023-11-22T13:28:21.025Z"
        }

        const component = renderer.create(
            <AppContext.Provider value={{user}}>
                <MemoryRouter>
                    <Header/>
                </MemoryRouter>
            </AppContext.Provider>);
        expect(component).toMatchSnapshot();
    })

    it("display the header without user information", () => {

        const user = null;


        const component = renderer.create(
            <AppContext.Provider value={{user}}>
                <MemoryRouter>
                    <Header/>
                </MemoryRouter>
            </AppContext.Provider>);
        expect(component).toMatchSnapshot();
    })


})