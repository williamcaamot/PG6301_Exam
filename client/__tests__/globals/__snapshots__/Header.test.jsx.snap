// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`header component display the header with user information 1`] = `
<div
  className="flexWrapper"
>
  <div
    className="headerWrapper"
  >
    <div
      style={
        {
          "padding": "30px",
        }
      }
    >
      <h1>
        ChatMania
      </h1>
    </div>
    <div>
      <nav>
        <a
          className="mainMenuLink"
          href="/"
          onClick={[Function]}
        >
          Hjem
        </a>
        <a
          className="mainMenuLink"
          href="/chat"
          onClick={[Function]}
        >
          Chat
        </a>
        <a
          className="mainMenuLink"
          href="/chatroom/add"
          onClick={[Function]}
        >
          Legg til nytt chat rom
        </a>
        <a
          className="mainMenuLink"
          href="/profile"
          onClick={[Function]}
        >
          Velkommen, 
          WILLIAM123wwwww
        </a>
      </nav>
    </div>
  </div>
</div>
`;

exports[`header component display the header without user information 1`] = `
<div
  className="flexWrapper"
>
  <div
    className="headerWrapper"
  >
    <div
      style={
        {
          "padding": "30px",
        }
      }
    >
      <h1>
        ChatMania
      </h1>
    </div>
    <div>
      <nav>
        <a
          className="mainMenuLink"
          href="/"
          onClick={[Function]}
        >
          Hjem
        </a>
        <a
          className="mainMenuLink"
          href="/chat"
          onClick={[Function]}
        >
          Chat
        </a>
        <a
          className="mainMenuLink"
          href="/chatroom/add"
          onClick={[Function]}
        >
          Legg til nytt chat rom
        </a>
        <a
          className="mainMenuLink"
          href="/login"
          onClick={[Function]}
        >
          Logg inn
        </a>
      </nav>
    </div>
  </div>
</div>
`;
